README.md
# Przeslijmi Sivalidator

[![Run Status](https://api.shippable.com/projects/5e4da2352dc4920007d91aa9/badge?branch=master)]()

[![Coverage Badge](https://api.shippable.com/projects/5e4da2352dc4920007d91aa9/coverageBadge?branch=master)]()



CHANGELOG.md
# Changelog

## v2.2.3 - 2020-04-15

- Updated `composer.json`.

## v2.2.2 - 2020-04-14

- Added `shippable.yml`.
- Added badges.
- Deleted `build.sh.dist`.

## v2.2.1 - 2019-11-10

- Added `build.sh.dist`.
- Added `phpcs.xml`.
- Added `phpunit.xml` to be more complete.
- Changed `composer.json`.

## v2.2.0 - 2019-10-25

- Added 100% code coverage report.

## v2.1.0 - 2019-09-12

- Added new class `Date`.
- Added Exception `DateFormatReturnUnknownException`.
- Added Exception `DateFormatWrongException`.
- Added Exception `DateFormatWrongObjectException`.
- Added Exception `DateMoveWrongSyntaxException`.
- Added all unit tests (code coverage = 100%).

## v2.0.0 - 2019-08-12

- Added all unit tests (code coverage = 100%).
- Added method `GeoProgression::has`.
- Changed name of method `GeoProgression::getProgression` to `GeoProgression::get`.
- Fixed `RegEx::ifNotMatches`.
- Fixed documentation.

## v1.0.0 - 2019-08-09

- Added all.


tests/DateTest.php
<?php declare(strict_types=1);

namespace Przeslijmi\Sivalidator;

use DateTime;
use PHPUnit\Framework\TestCase;
use Przeslijmi\Sivalidator\Date;
use Przeslijmi\Sivalidator\Exceptions\DateFormatReturnUnknownException;
use Przeslijmi\Sivalidator\Exceptions\DateFormatWrongException;
use Przeslijmi\Sivalidator\Exceptions\DateFormatWrongObjectException;
use Przeslijmi\Sivalidator\Exceptions\DateMoveWrongSyntaxException;
use stdClass;

/**
 * Methods for testing Date.
 */
final class DateTest extends TestCase
{

    /**
     * Provides various date formats to test.
     *
     * @return array[][]
     */
    public function datesProvider() : array
    {

        // Lvd.
        $dtObj = new DateTime('2019-01-01');

        return [
            [ $dtObj,       '2019-01-01', 'DateTimeObject' ],
            [ '2019',       '2019-01-01', 'YearPeriod'     ],
            [ '2019H1',     '2019-01-01', 'HalfYearPeriod' ],
            [ '2019H2',     '2019-07-01', 'HalfYearPeriod' ],
            [ '2019Q1',     '2019-01-01', 'QuarterPeriod'  ],
            [ '2019Q2',     '2019-04-01', 'QuarterPeriod'  ],
            [ '2019M01',    '2019-01-01', 'MonthPeriod'    ],
            [ '2019M02',    '2019-02-01', 'MonthPeriod'    ],
            [ '2019W01',    '2019-01-01', 'WeekPeriod'     ],
            [ '2019W02',    '2019-01-08', 'WeekPeriod'     ],
            [ '2019D001',   '2019-01-01', 'DatePeriod'     ],
            [ '2019D002',   '2019-01-02', 'DatePeriod'     ],
            [ '2019-01-01', '2019-01-01', 'YmdDate'        ],
        ];
    }

    /**
     * Provides various date formats to test.
     *
     * @return array[][]
     */
    public function movesProvider() : array
    {

        return [
            [ '1Y', '2020-01-01' ],
            [ '1H', '2019-07-01' ],
            [ '1Q', '2019-04-01' ],
            [ '1M', '2019-02-01' ],
            [ '1W', '2019-01-08' ],
            [ '1D', '2019-01-02' ],
            [ '2Y', '2021-01-01' ],
            [ '2H', '2020-01-01' ],
            [ '2Q', '2019-07-01' ],
            [ '2M', '2019-03-01' ],
            [ '2W', '2019-01-15' ],
            [ '2D', '2019-01-03' ],
            [ '-1Y', '2018-01-01' ],
            [ '-1H', '2018-07-01' ],
            [ '-1Q', '2018-10-01' ],
            [ '-1M', '2018-12-01' ],
            [ '-1W', '2018-12-25' ],
            [ '-1D', '2018-12-31' ],
            [ '-2Y', '2017-01-01' ],
            [ '-2H', '2018-01-01' ],
            [ '-2Q', '2018-07-01' ],
            [ '-2M', '2018-11-01' ],
            [ '-2W', '2018-12-18' ],
            [ '-2D', '2018-12-30' ],
        ];
    }

    /**
     * Provides various date formats to test.
     *
     * @return array[][]
     */
    public function returnsProvider() : array
    {

        // Lvd.
        $dtObj = new DateTime('2020-01-01');

        return [
            [ '2020-01-01', null,             '2020-01-01' ],
            [ '2020-01-01', 'YmdDate',        '2020-01-01' ],
            [ '2020-01-01', 'YearPeriod',     '2020'       ],
            [ '2020-01-01', 'HalfYearPeriod', '2020H1'     ],
            [ '2020-01-01', 'QuarterPeriod',  '2020Q1'     ],
            [ '2020-01-01', 'MonthPeriod',    '2020M01'    ],
            [ '2020-01-01', 'WeekPeriod',     '2020W01'    ],
            [ '2020-01-01', 'DayPeriod',      '2020D001'   ],
            [ '2020-01-01', 'DateTimeObject', $dtObj       ],
        ];
    }

    /**
     * Provides various wrong periods as strings.
     *
     * @return array[][]
     */
    public function wrongStringProvider() : array
    {

        return [
            [ '1'        ],
            [ '2019H0'   ],
            [ '2019H3'   ],
            [ '2019Q0'   ],
            [ '2019Q5'   ],
            [ '2019Q03'  ],
            [ '2019M00'  ],
            [ '2019M13'  ],
            [ '2019D467' ],
            [ '2019D0'   ],
            [ '2019D00'  ],
            [ '2019D1'   ],
        ];
    }

    /**
     * Test if creation works.
     *
     * @param DataTime|string $period    Period to test.
     * @param string          $startDate Calculated proper start date.
     * @param string          $format    Calculated proper format of `$period`.
     *
     * @return void
     *
     * @dataProvider datesProvider
     */
    public function testCreation($period, string $startDate, string $format) : void
    {

        // Create.
        $date = new Date($period);

        // Test.
        $this->assertEquals($startDate, $date->get());
        $this->assertEquals($format, $date->getFormat());
    }

    /**
     * Test if moving works.
     *
     * @param string $move         What move to make from 2019-01-01.
     * @param string $expectedDate What date is to be expected.
     *
     * @return void
     *
     * @dataProvider movesProvider
     */
    public function testMoving(string $move, string $expectedDate) : void
    {

        // Create.
        $date = new Date('2019');
        $date->move($move);

        // Test.
        $this->assertEquals($expectedDate, $date->get());
    }

    /**
     * Test if moving works.
     *
     * @param string          $dateToTest     Date (Y-m-d) to be tested.
     * @param null|string     $returnFormat   What format to call.
     * @param DataTime|string $expectedReturn Expected result of transferring date to format.
     *
     * @return void
     *
     * @dataProvider returnsProvider
     */
    public function testReturning(string $dateToTest, $returnFormat, $expectedReturn) : void
    {

        // Lvd.
        $date = new Date($dateToTest);

        // Test.
        if (is_null($returnFormat) === true) {
            $this->assertEquals($expectedReturn, $date->get());
        } else {
            $this->assertEquals($expectedReturn, $date->get($returnFormat));
        }
    }

    /**
     * Test if sending wrong object to class throws.
     *
     * @return void
     */
    public function testIfSendingWrongObjectThrows() : void
    {

        // Prepare.
        $this->expectException(DateFormatWrongObjectException::class);

        // Test.
        new Date(new stdClass());
    }

    /**
     * Test if sending wrong datatype to class throws.
     *
     * @return void
     */
    public function testIfSendingWrongDataTypeThrows() : void
    {

        // Prepare.
        $this->expectException(DateFormatWrongObjectException::class);

        // Test.
        new Date(true);
    }

    /**
     * Test if sending string wrong syntaxed period throws.
     *
     * @param string $wrongPeriod Wrong period to be tested.
     *
     * @return void
     *
     * @dataProvider wrongStringProvider
     */
    public function testIfSendingWrongSyntaxedPeriodThrows(string $wrongPeriod) : void
    {

        // Prepare.
        $this->expectException(DateFormatWrongException::class);

        // Test.
        new Date($wrongPeriod);
    }

    /**
     * Test if sending string wrong syntaxed period throws.
     *
     * @return void
     */
    public function testIfMovingWithWrongUnitThrows() : void
    {

        // Prepare.
        $this->expectException(DateMoveWrongSyntaxException::class);

        // Test.
        $date = new Date('2019');
        $date->move('1T');
    }

    /**
     * Test if asking for wrong return format throws.
     *
     * @return void
     */
    public function testIfAskingForWrongReturnFormatThrows() : void
    {

        // Prepare.
        $this->expectException(DateFormatReturnUnknownException::class);

        // Test.
        $date = new Date('2019');
        $date->get('NonexistingPeriod');
    }
}


tests/RegExTest.php
<?php declare(strict_types=1);

namespace Przeslijmi\Sivalidator;

use PHPUnit\Framework\TestCase;
use Przeslijmi\Sexceptions\Exceptions\ParamWrosynException;
use Przeslijmi\Sexceptions\Exceptions\RegexTestFailException;
use Przeslijmi\Sivalidator\RegEx;
use stdClass;
use TypeError;

/**
 * Methods for testing values against regex syntax.
 */
final class RegExTest extends TestCase
{

    /**
     * Test proper string and pattern and assert true.
     *
     * @return void
     */
    public function testIfHelloMatchesTrue1() : void
    {

        $this->assertTrue(Regex::ifMatches('Hello', '/^([A-Za-z]+)$/', false));
        $this->assertFalse(Regex::ifNotMatches('Hello', '/^([A-Za-z]+)$/', false));
    }

    /**
     * Test proper string and pattern and assert true but leave third optional parameter.
     *
     * @return void
     */
    public function testIfHelloMatchesTrue2() : void
    {

        $this->assertTrue(Regex::ifMatches('Hello', '/^([A-Za-z]+)$/'));
        $this->assertFalse(Regex::ifNotMatches('Hello', '/^([A-Za-z]+)$/', false));
    }

    /**
     * Test inproper string and pattern and assert false.
     *
     * @return void
     */
    public function testIfHelloMatchesFalse1() : void
    {

        $this->assertFalse(
            Regex::ifMatches('Hello 123', '/^([A-Za-z]+)$/', false),
            'Hello 123 should not match because it has digits.'
        );
        $this->assertTrue(Regex::ifNotMatches('Hello 123', '/^([A-Za-z]+)$/', false));
    }

    /**
     * Test inproper string and pattern will throw exception.
     *
     * @return void
     */
    public function testIfHelloThrows1() : void
    {

        $this->expectException(RegexTestFailException::class);

        Regex::ifMatches('Hello 123', '/^([A-Za-z]+)$/');
    }

    /**
     * Test proper string and pattern will throw exception.
     *
     * @return void
     */
    public function testIfHelloThrows2() : void
    {

        $this->expectException(RegexTestFailException::class);

        Regex::ifNotMatches('Hello 123', '/^([A-Za-z0-9 ]+)$/');
    }

    /**
     * Test inproper type of first param of ifMatches throws.
     *
     * @return void
     */
    public function testIfWrotypeThrows1() : void
    {

        $this->expectException(TypeError::class);

        Regex::ifMatches(new stdClass(), '/^([A-Za-z]+)$/');
    }

    /**
     * Test inproper type of second param of ifMatches throws.
     *
     * @return void
     */
    public function testIfWrotypeThrows2() : void
    {

        $this->expectException(TypeError::class);

        Regex::ifMatches('aaa', true);
    }

    /**
     * Test nonregex string in second param of ifMatches throws.
     *
     * @return void
     */
    public function testIfWrotypeThrows3() : void
    {

        $this->expectException(ParamWrosynException::class);

        Regex::ifMatches('aaa', 'WRONG_REGEX_STRING');
    }

    /**
     * Test nonregex string in second param of ifMatches throws.
     *
     * @return void
     */
    public function testIfWrotypeThrows4() : void
    {

        $this->expectException(ParamWrosynException::class);

        Regex::ifMatches('aaa', '/');
    }
}


tests/TypeHintingTest.php
<?php declare(strict_types=1);

namespace Przeslijmi\Sivalidator;

use PHPUnit\Framework\TestCase;
use Przeslijmi\Sexceptions\Exceptions\TypeHintingFailException;
use stdClass;

/**
 * Methods for testing values against type hinting class.
 */
final class TypeHintingTest extends TestCase
{

    /**
     * Test if "isArrayOf" returns true on positive values.
     *
     * @return void
     */
    public function testIfIsArrayOfRtot() : void
    {

        $array = [
            ( new stdClass() ),
            ( new stdClass() ),
            ( new stdClass() ),
        ];

        $this->assertTrue(TypeHinting::isArrayOf($array, 'stdClass'));
    }

    /**
     * Test if "isArrayOf" returns false on negative values.
     *
     * @return void
     */
    public function testIfIsArrayOfRfof1() : void
    {

        $array = [
            ( new stdClass() ),
            ( new stdClass() ),
            ( new stdClass() ),
        ];

        $this->assertFalse(TypeHinting::isArrayOf($array, 'Other\Name', false));
    }

    /**
     * Test if "isArrayOf" returns false on negative values.
     *
     * @return void
     */
    public function testIfIsArrayOfRfof2() : void
    {

        $array = [
            'string',
            'string',
            'string',
        ];

        $this->assertFalse(TypeHinting::isArrayOf($array, 'Other\Name', false));
    }

    /**
     * Test if "isArrayOf" throws.
     *
     * @return void
     */
    public function testIfIsArrayOfThrows() : void
    {

        $this->expectException(TypeHintingFailException::class);

        $array = [
            ( new stdClass() ),
            ( new stdClass() ),
            ( new stdClass() ),
        ];

        $this->assertFalse(TypeHinting::isArrayOf($array, 'Other\Name'));
    }

    /**
     * Test if "isArrayOfStrings" returns true on positive values.
     *
     * @return void
     */
    public function testIfIsArrayOfStringsRtot() : void
    {

        $array = [ 'string1', 'string2', 'string3' ];

        $this->assertTrue(TypeHinting::isArrayOfStrings($array));
    }

    /**
     * Test if "isArrayOfStrings" returns true on positive values (including nulls).
     *
     * @return void
     */
    public function testIfIsArrayOfStringsRtotWithNulls() : void
    {

        $array = [ 'string1', 'string2', 'string3', null ];

        $this->assertTrue(TypeHinting::isArrayOfStrings($array, true));
    }

    /**
     * Test if "isArrayOfStrings" returns false on negative values.
     *
     * @return void
     */
    public function testIfIsArrayOfStringsRfof() : void
    {

        $array = [ 'string1', 3, true ];

        $this->assertFalse(TypeHinting::isArrayOfStrings($array, false, false));
    }

    /**
     * Test if "isArrayOfStrings" returns true on positive values (including nulls).
     *
     * @return void
     */
    public function testIfIsArrayOfStringsRfofWithNulls() : void
    {

        $array = [ 'string1', 'string2', 'string3', null ];

        $this->assertTrue(TypeHinting::isArrayOfStrings($array, true, false));
    }

    /**
     * Test if "isArrayOfStrings" throws.
     *
     * @return void
     */
    public function testIfIsArrayOfStringsThrows() : void
    {

        $this->expectException(TypeHintingFailException::class);

        $array = [ 'string1', 3, true ];

        $this->assertFalse(TypeHinting::isArrayOfStrings($array));
    }

    /**
     * Test if "isArrayOfScalars" returns true on positive values.
     *
     * @return void
     */
    public function testIfIsArrayOfScalarsRtot() : void
    {

        $array = [ 5, true, 1.1, 'string' ];

        $this->assertTrue(TypeHinting::isArrayOfScalars($array));
    }

    /**
     * Test if "isArrayOfScalars" returns false on negative values.
     *
     * @return void
     */
    public function testIfIsArrayOfScalarsRfof() : void
    {

        $array = [ [ 5 , true ], 1.1, 'string' ];

        $this->assertFalse(TypeHinting::isArrayOfScalars($array, false));
    }

    /**
     * Test if "isArrayOfScalars" throws.
     *
     * @return void
     */
    public function testIfIsArrayOfScalarsThrows() : void
    {

        $this->expectException(TypeHintingFailException::class);

        $array = [ [ 5 , true ], 1.1, 'string' ];

        $this->assertFalse(TypeHinting::isArrayOfScalars($array));
    }
}


src/RegEx.php
<?php declare(strict_types=1);

namespace Przeslijmi\Sivalidator;

use Przeslijmi\Sexceptions\Exceptions\RegexTestFailException;
use Przeslijmi\Sexceptions\Exceptions\ParamWrosynException;

/**
 * Methods for testing values against regex syntax.
 */
class RegEx
{

    /**
     * Tests if string matches given regex (can return RegexTestFailException).
     *
     * ## Usage example
     * ```
     * RegEx::ifMatches('Hello', '/^([a-z]+)$/');        // will throw RegexTestFailException
     * RegEx::ifMatches('Hello', '/^([a-z]+)$/', false); // will return false
     * RegEx::ifMatches('Hello', '/^([a-zA-Z]+)$/');     // will return true
     * ```
     *
     * @param string  $what  String to search in.
     * @param string  $regex Regular expression.
     * @param boolean $throw Opt., true. Set to false to prevent throwing.
     *
     * @since  v1.0
     * @throws RegexTestFailException When RegEx will return false, and throwing is 'On'.
     * @throws ParamWrosynException   When second param (regex) is corrupted.
     * @throws ParamWrosynException   When second param (regex) is too short (<=2).
     * @return boolean
     */
    public static function ifMatches(string $what, string $regex, bool $throw = true) : bool
    {

        // Check first and last letter of $regex.
        $testRegex = rtrim($regex, 'imsxuADU');
        if (substr($testRegex, 0, 1) !== '/' || substr($testRegex, -1) !== '/') {
            $hint  = 'Regex has to start and and with a slash.';
            $hint .= 'Haven\'t you replaced order of params in method call?';
            throw (new ParamWrosynException('regex', $regex))->addInfo('hint', $hint);
        }

        // Check regex min length.
        if (strlen($regex) < 2) {
            throw (new ParamWrosynException('regex', $regex))
                ->addInfo('hint', 'Regex has to be at least 2 char length.');
        }

        // Lvd & test.
        $matches = (array) preg_grep($regex, [ $what ]);
        $test    = (bool) count($matches);

        // Throw.
        if ($throw === true && $test === false) {
            throw new RegexTestFailException($what, $regex);
        }

        return $test;
    }

    /**
     * Tests if string NOT matches given regex (can return RegexTestFailException).
     *
     * ## Usage example
     * ```
     * RegEx::ifNotMatches('Hello', '/^([a-z]+)$/');           // will return true
     * RegEx::ifNotMatches('Hello', '/^([a-z]+)$/', false);    // will return true
     * RegEx::ifNotMatches('Hello', '/^([a-zA-Z]+)$/');        // will throw RegexTestFailException
     * RegEx::ifNotMatches('Hello', '/^([a-zA-Z]+)$/', false); // will return false
     * ```
     *
     * @param string  $what  String to search in.
     * @param string  $regex Regular expression.
     * @param boolean $throw Opt., true. Set to false to prevent throwing.
     *
     * @throws   RegexTestFailException When RegEx will return true, and throwing is 'On'.
     * @rethrows ParamWrosynException   When second param (regex) is corrupted.
     * @since    v1.0
     * @return   boolean
     */
    public static function ifNotMatches(string $what, string $regex, bool $throw = true) : bool
    {

        // Get test.
        $test = self::ifMatches($what, $regex, false);

        // Reverse it.
        $test = ( ! $test );

        // Serve excception.
        if ($throw === true && $test === false) {
            throw new RegexTestFailException($what, 'OPPOSITE TO: ' . $regex);
        }

        return $test;
    }
}


src/TypeHinting.php
<?php declare(strict_types=1);

namespace Przeslijmi\Sivalidator;

use Przeslijmi\Sexceptions\Exceptions\TypeHintingFailException;

/**
 * Methods for testing values against hint types.
 */
class TypeHinting
{

    /**
     * Tests if array consists of elements given type.
     *
     * ## Usage example
     * ```
     * $array = [ (new \stdClass()), (new \stdClass()), (new \stdClass()) ];
     * TypeHinting::isArrayOf($array, 'Namespace\Class');
     * ```
     *
     * @param array   $arrayOfObjects Array to be checked.
     * @param string  $className      Name of the class that is expected.
     * @param boolean $throw          Opt., true. Set to false to prevent throwing.
     *
     * @since  v1.0
     * @throws TypeHintingFailException If this is not array of given objects.
     * @return boolean
     */
    public static function isArrayOf(array $arrayOfObjects, string $className, bool $throw = true) : bool
    {

        // Lvd.
        $test = true;
        $isa  = '';

        // Test.
        foreach ($arrayOfObjects as $object) {

            if (is_object($object) === false) {
                $test = false;
                $isa  = gettype($object) . '[]';
                break;
            }

            if (is_a($object, $className) === false) {
                $test = false;
                $isa  = get_class($object) . '[]';
                break;
            }
        }

        // Throw.
        if ($throw === true && $test === false) {
            throw new TypeHintingFailException($className . '[]', $isa);
        }

        return $test;
    }

    /**
     * Tests if array consists of elements given type.
     *
     * ## Usage example
     * ```
     * $array = [ 'string1', 'string2', 'string3' ];
     * TypeHinting::isArrayOfStrings($array);
     * ```
     *
     * @param array   $arrayOfObjects Array to be checked.
     * @param boolean $acceptNulls    Opt., false. Set to true to also accept null values.
     * @param boolean $throw          Opt., true. Set to false to prevent throwing.
     *
     * @throws TypeHintingFailException If this is not array of strings.
     * @since  v1.0
     * @return boolean
     */
    public static function isArrayOfStrings(array $arrayOfObjects, bool $acceptNulls = false, bool $throw = true) : bool
    {

        // Lvd.
        $test = true;
        $isa  = '';

        // Test.
        foreach ($arrayOfObjects as $object) {

            $isProblem = ( is_string($object) === false );

            if ($object === null && $acceptNulls === true) {
                $isProblem = false;
            }

            if ($isProblem === true) {
                $test = false;
                $isa  = gettype($object) . '[]';
                break;
            }
        }

        // Throw.
        if ($throw === true && $test === false) {
            throw new TypeHintingFailException('string[]', $isa);
        }

        return $test;
    }

    /**
     * Tests if array consists of elements given type.
     *
     * ## Usage example
     * ```
     * $array = [ 5, true, 1.1, 'string' ];
     * TypeHinting::isArrayOfScalars($array);
     * ```
     *
     * @param array   $arrayOfObjects Array to be checked.
     * @param boolean $throw          Opt., true. Set to false to prevent throwing.
     *
     * @throws TypeHintingFailException If this is not array of scalars.
     * @since  v1.0
     * @return boolean
     */
    public static function isArrayOfScalars(array $arrayOfObjects, bool $throw = true) : bool
    {

        // Lvd.
        $test = true;
        $isa  = '';

        // Test.
        foreach ($arrayOfObjects as $object) {
            if (is_scalar($object) === false) {
                $test = false;
                $isa  = gettype($object) . '[]';
                break;
            }
        }

        // Throw.
        if ($throw === true && $test === false) {
            throw new TypeHintingFailException('scalar[]', $isa);
        }

        return $test;
    }
}


src/Exceptions/DateFormatReturnUnknownException.php
<?php declare(strict_types=1);

namespace Przeslijmi\Sivalidator\Exceptions;

use Przeslijmi\Sexceptions\Exceptions\MethodFopException;

/**
 * Date format for return is corrupted.
 */
class DateFormatReturnUnknownException extends MethodFopException
{

    /**
     * Constructor.
     *
     * @param string $returnFormat Return format sent to use.
     *
     * @since v1.0
     *
     * phpcs:disable Generic.Files.LineLength
     */
    public function __construct(string $returnFormat)
    {

        $this->addInfo('context', 'returningDataObject');
        $this->addInfo('returnFormat', $returnFormat);
        $this->addInfo('hint', 'Possible return formats are DateTimeObject, YmdDate, YearPeriod, HalfYearPeriod, QuarterPeriod, MonthPeriod, WeekPeriod, DayPeriod. Other return format was sent.');
    }
}


src/Date.php
<?php declare(strict_types=1);

namespace Przeslijmi\Sivalidator;

use DateInterval;
use DateTime;
use Przeslijmi\Sivalidator\Exceptions\DateFormatReturnUnknownException;
use Przeslijmi\Sivalidator\Exceptions\DateFormatWrongException;
use Przeslijmi\Sivalidator\Exceptions\DateFormatWrongObjectException;
use Przeslijmi\Sivalidator\Exceptions\DateMoveWrongSyntaxException;
use Przeslijmi\Sivalidator\RegEx;

/**
 * Date counting solution.
 *
 * ## Usage example
 * ```
 * $date = new \Przeslijmi\Sivalidator\Date('2019H2');
 * $date->move('1Y2W-2D');
 * var_dump($date->get()); // will return 2022-07-13
 * ```
 */
class Date
{

    /**
     * Date as string.
     *
     * @var string
     */
    private $date;

    /**
     * What was the original format of given date.
     *
     * Possible options:
     *   - DateTimeObject
     *   - YearPeriod
     *   - HalfYearPeriod
     *   - QuarterPeriod
     *   - MonthPeriod
     *   - WeekPeriod
     *   - DayPeriod
     *   - YmdDate
     *
     * @var string
     */
    private $format;

    /**
     * Constructor.
     *
     * @param DateTime|string $date Date as DateTime object or string.
     *
     * @since v1.0
     */
    public function __construct($date)
    {

        $this->setDate($date);
    }

    /**
     * Setter (and validator) for date.
     *
     * @param DateTime|string $date Date as DateTime object or string.
     *
     * @return self
     * @throws DateFormatWrongObjectException When format is not DateTime, nor string.
     * @throws DateFormatWrongException       When format is string but unrecognizable.
     * @since  v1.0
     *
     * @phpcs:disable Generic.Metrics.CyclomaticComplexity
     */
    private function setDate($date) : self
    {

        // This is a DateTime object.
        if (is_a($date, 'DateTime') === true) {

            // Save.
            $this->date   = $date->format('Y-m-d');
            $this->format = 'DateTimeObject';

            return $this;
        }

        // This is not a string, nor a DateTime - throw.
        if (is_object($date) === true) {
            throw new DateFormatWrongObjectException(get_class($date));
        } elseif (is_string($date) === false) {
            throw new DateFormatWrongObjectException(gettype($date));
        }

        // Lvd.
        $monthRegex = '/^([0-9]){4}(M)((01)|(02)|(03)|(04)|(05)|(06)|(07)|(08)|(09)|(10)|(11)|(12))$/';

        // Find proper date if string.
        if (RegEx::ifMatches($date, '/^([0-9]){4}$/', false) === true) {

            // Define.
            $date   = ( new DateTime($date . '-01-01') );
            $format = 'YearPeriod';

        } elseif (RegEx::ifMatches($date, '/^([0-9]){4}(H)([12])$/', false) === true) {

            // Lvd.
            list($year, $halfyear) = explode('H', $date);

            // Define.
            $date   = ( new DateTime($year . '-' . ( ( ( (int) $halfyear - 1 ) * 6 ) + 1 ) . '-01') );
            $format = 'HalfYearPeriod';

        } elseif (RegEx::ifMatches($date, '/^([0-9]){4}(Q)([1234])$/', false) === true) {

            // Lvd.
            list($year, $quarter) = explode('Q', $date);

            // Define.
            $date   = ( new DateTime($year . '-' . ( ( ( (int) $quarter - 1 ) * 3 ) + 1 ) . '-01') );
            $format = 'QuarterPeriod';

        } elseif (RegEx::ifMatches($date, $monthRegex, false) === true) {

            // Lvd.
            list($year, $month) = explode('M', $date);

            // Define.
            $date   = ( new DateTime($year . '-' . $month . '-01') );
            $format = 'MonthPeriod';

        } elseif (RegEx::ifMatches($date, '/^([0-9]){4}(W)([012345])([0-9])$/', false) === true) {

            // Lvd.
            list($year, $week) = explode('W', $date);
            $days              = ( ( ( (int) $week ) - 1 ) * 7 );

            // Define.
            $date   = ( new DateTime($year . '-01-01') )->add(new DateInterval('P' . $days . 'D'));
            $format = 'WeekPeriod';

        } elseif (RegEx::ifMatches($date, '/^(20)([12])([0-9])(D)([0123])([0-9])([0-9])$/', false) === true) {

            // Lvd.
            list($year, $day) = explode('D', $date);
            $days             = ( ( (int) $day ) - 1 );

            // Define.
            $date   = ( new DateTime($year . '-01-01') )->add(new DateInterval('P' . $days . 'D'));
            $format = 'DatePeriod';

        } elseif (RegEx::ifMatches($date, '/^([0-9]){4}-([0-9]){2}-([0-9]){2}$/', false) === true) {

            // Lvd.
            list($year, $month, $day) = explode('-', $date);

            // Define.
            $date   = new DateTime($year . '-' . $month . '-' . $day);
            $format = 'YmdDate';

        } else {

            // Nothing fits - throw.
            throw new DateFormatWrongException($date);
        }//end if

        // Save date & format.
        $this->date   = $date->format('Y-m-d');
        $this->format = $format;

        return $this;
    }

    /**
     * Move date with given move description.
     *
     * ## Usage example
     * ```
     * $date = ( new Date('2019') )->move('1Y');
     * ```
     *
     * ## Possible units
     * - `Y` - for year
     * - `H` - for halfyear
     * - `Q` - for quarter
     * - `M` - for month
     * - `W` - for week
     * - `D` - for day
     *
     * @param string $move Move description.
     *
     * @throws DateMoveWrongSyntaxException When move order has wrong syntax.
     * @return self
     * @since  v1.0
     */
    public function move(string $move) : self
    {

        // Lvd.
        $moveMonths = 0;
        $moveDays   = 0;

        // Find moves.
        preg_match_all('/(-)?([0-9])+([YHQMWD]){1}/', $move, $foundMoves);
        $foundMoves = $foundMoves[0];

        // Throw.
        if (count($foundMoves) === 0 && empty($move) === false) {
            throw new DateMoveWrongSyntaxException($move);
        }

        // Serve every move.
        foreach ($foundMoves as $move) {

            // Lvd.
            // Amount can be negative.
            $amount = (int) substr($move, 0, -1);
            $unit   = (string) substr($move, -1);

            // Get starting date.
            list($year, $month, $day) = explode('-', $this->date);

            // Integerize.
            $month = (int) $month;
            $day   = (int) $day;
            $year  = (int) $year;

            // Recalc move.
            switch ($unit) {
                case 'Y':
                    $moveMonths = ( 12 * $amount );
                break;

                case 'H':
                    $moveMonths = ( 6 * $amount );
                break;

                case 'Q':
                    $moveMonths = ( 3 * $amount );
                break;

                case 'M':
                    $moveMonths = $amount;
                break;

                case 'W':
                    $moveDays = ( 7 * $amount );
                break;

                case 'D':
                default:
                    $moveDays = $amount;
                break;
            }//end switch

            // Make move.
            $this->date = date('Y-m-d', mktime(0, 0, 0, ( $month + $moveMonths ), ( $day + $moveDays ), $year));

        }//end foreach

        return $this;
    }

    /**
     * Get date as Y-m-d string.
     *
     * @param string $format Format name to use on return, ie. DateTimeObject, YmdDate,
     *                       YearPeriod, HalfYearPeriod, QuarterPeriod, MonthPeriod, WeekPeriod,
     *                       DayPeriod.
     *
     * @throws DateFormatReturnUnknownException When unknown format has been sent.
     * @return DateTime|string
     * @since  v1.0
     */
    public function get(string $format = 'YmdDate')
    {

        // Lvd.
        list($year, $month, $day) = explode('-', $this->date);

        // Integerize.
        $month = (int) $month;
        $day   = (int) $day;
        $year  = (int) $year;

        // Deliver depending on format.
        switch ($format) {

            case 'DateTimeObject':
            return new DateTime($this->date);

            case 'YmdDate':
            return $this->date;

            case 'YearPeriod':
            return (string) $year;

            case 'HalfYearPeriod':
            return (string) $year . 'H' . ceil(( $month / 6 ));

            case 'QuarterPeriod':
            return (string) $year . 'Q' . ceil(( $month / 3 ));

            case 'MonthPeriod':
            return (string) $year . 'M' . str_pad((string) $month, 2, '0', STR_PAD_LEFT);

            case 'WeekPeriod':
                $weekNo = (int) ( new DateTime($this->date) )->format('W');
            return (string) $year . 'W' . str_pad((string) $weekNo, 2, '0', STR_PAD_LEFT);

            case 'DayPeriod':
                $weekNo = (int) ( ( new DateTime($this->date) )->format('z') + 1 );
            return (string) $year . 'D' . str_pad((string) $weekNo, 3, '0', STR_PAD_LEFT);

            default:
            throw new DateFormatReturnUnknownException($format);
        }//end switch
    }

    /**
     * Getter for format.
     *
     * @since  v1.0
     * @return string
     */
    public function getFormat() : string
    {

        return $this->format;
    }
}
